src/MerchantCenter/MerchantStatuses.php

```diff
@@ -251,53 +251,67 @@ protected function fetch_issues( string $type = null, int $per_page = 0, int $pa
+		$merchant = $this->container->get( Merchant::class );
 		/** @var ProductHelper $product_helper */
-		$product_helper = $this->container->get( ProductHelper::class );
+		$product_helper      = $this->container->get( ProductHelper::class );
+		$visibility_meta_key = $this->prefix_meta_key( ProductMetaHandler::KEY_VISIBILITY );
 
-		return array_values(
-			array_filter(
-				$mc_statuses,
-				function( $product ) use ( $product_helper ) {
-					$wc_product_id = $product_helper->get_wc_product_id( $product->getProductId() );
-					// Skip products not synced by this extension.
-					if ( ! $wc_product_id ) {
-						return false;
-					}
+		$valid_statuses = [];
+		foreach ( $merchant->get_productstatuses_batch( $google_ids )->getEntries() as $response_entry ) {
+			$mc_product_status = $response_entry->getProductStatus();
+			if ( ! $mc_product_status ) {
+				do_action(
+					'gla_debug_message',
+					sprintf( 'A Google ID was not found in Merchant Center.' ),
+					__METHOD__
+				);
+				continue;
+			}
+			$mc_product_id = $mc_product_status->getProductId();
+			$wc_product_id = $product_helper->get_wc_product_id( $mc_product_id );
+			// Skip products not synced by this extension.
+			if ( ! $wc_product_id ) {
+				continue;
+			}
 
-					// Product previously found/validated.
-					if ( ! empty( $this->product_data_lookup[ $wc_product_id ] ) ) {
-						return true;
-					}
+			// Product previously found/validated.
+			if ( ! empty( $this->product_data_lookup[ $wc_product_id ] ) ) {
+				$valid_statuses[] = $response_entry->getProductStatus();
+				continue;
+			}
 
-					$wc_product = wc_get_product( $wc_product_id );
-					// Skip products that are no longer in WooCommerce.
-					if ( ! $wc_product ) {
-						do_action(
-							'woocommerce_gla_debug_message',
-							sprintf( 'Merchant Center product %s not found in this WooCommerce store.', $product->getProductId() ),
-							__METHOD__ . ' in remove_invalid_statuses()',
-						);
-						return false;
-					}
+			$wc_product = get_post( $wc_product_id );
+			if ( ! $wc_product || 'product' !== substr( $wc_product->post_type, 0, 7 ) ) {
+				// Should never reach here since the products IDS are retrieved from postmeta.
+				do_action(
+					'woocommerce_gla_debug_message',
+					sprintf( 'Merchant Center product %s not found in this WooCommerce store.', $mc_product_id ),
+					__METHOD__ . ' in remove_invalid_statuses()',
+				);
+				continue;
+			}
 
-					$this->product_data_lookup[ $wc_product_id ] = [
-						'name'       => $wc_product->get_name(),
-						'visibility' => $wc_product->get_meta( $this->prefix_meta_key( ProductMetaHandler::KEY_VISIBILITY ) ),
-					];
-					return true;
-				}
-			)
-		);
+			$this->product_data_lookup[ $wc_product_id ] = [
+				'name'            => get_the_title( $wc_product ),
+				'visibility'      => get_post_meta( $wc_product_id, $visibility_meta_key ),
+				'maybe_parent_id' => $wc_product->post_parent ?: $wc_product_id,
```

Comment:
Now that we added a lookup, doesn't it make more sense to just store `parent` here even if it could be null. Then later on we can just compare to see if has a value to determine if it's a child.

Reply:
A bit easier 03c8a4a1d3195840f065bd2ac00d4f1676fab2ce